packages/myst-execute/src/execute.ts

```diff
@@ -0,0 +1,340 @@
+}
+
+/**
+ * For each executable node, perform a kernel execution request and return the results.
+ * Return an additional boolean indicating whether an error occured.
+ *
+ * @param kernel
+ * @param nodes
+ * @param vfile
+ */
+async function computeExecutableNodes(
+  kernel: Kernel.IKernelConnection,
+  nodes: (ICellBlock | InlineExpression)[],
+  vfile: VFile,
+): Promise<{
+  results: (IOutput[] | IExpressionResult)[];
+  errorOccurred: boolean;
+}> {
+  let errorOccurred = false;
+
+  const results: (IOutput[] | IExpressionResult)[] = [];
+  for (const matchedNode of nodes) {
+    if (isCellBlock(matchedNode)) {
+      // Pull out code to execute
+      const code = select('code', matchedNode) as Code;
+      const { status, outputs } = await executeCode(kernel, code.value);
+      // Cache result
+      results.push(outputs);
+
+      // Check for errors
+      const metadata = matchedNode.data || {};
+      const allowErrors = !!metadata?.tags?.['raises-exception'];
+      if (status === 'error' && !allowErrors) {
+        fileWarn(vfile, 'An exception occurred during code execution, halting further execution');
+        // Make a note of the failure
+        errorOccurred = true;
+        break;
+      }
+    } else if (isInlineExpression(matchedNode)) {
+      // Directly evaluate the expression
+      const { status, result } = await evaluateExpression(kernel, matchedNode.value);
+
+      // Check for errors
+      if (status === 'error') {
+        fileWarn(
+          vfile,
+          'An exception occurred during expression evaluation, halting further execution',
+        );
+        // Make a note of the failure
+        errorOccurred = true;
+        break;
+      }
+      // Cache result
+      results.push(result);
+    } else {
+      assert(false);
+    }
+  }
+  return { results, errorOccurred };
+}
+
+/**
+ * Apply computed outputs (MIME bundles, stdout, etc.) to MDAST.
+ * If an unanticipated error occurred during execution, the length of
+ * `computedResult` may not correspond to that of `nodes` (it may be shorter)
+ *
+ * @param nodes executable MDAST nodes
+ * @param computedResult computed results for each node
+ */
+function applyComputedOutputsToNodes(
+  nodes: (ICellBlock | InlineExpression)[],
+  computedResult: (IOutput[] | IExpressionResult)[],
+) {
+  for (const matchedNode of nodes) {
+    // Pull out the result for this node
+    const thisResult = computedResult.shift();
+
+    if (isCellBlock(matchedNode)) {
+      // Pull out output to set data
+      const output = select('output', matchedNode) as unknown as { data: IOutput[] };
+      // Set the output array to empty if we don't have a result (e.g. due to a kernel error)
+      output.data = thisResult === undefined ? [] : (thisResult as IOutput[]);
+    } else if (isInlineExpression(matchedNode)) {
+      // Set data of expression to the result, or empty if we don't have one
+      matchedNode.result = // TODO: FIXME .data
```

Comment:
N.B. this should be `.data`, but will require a change to `myst-spec` too.

Reply:
We could target that change in another PR? It will also require a change in `jupyterlab-myst` and `myst-theme`? This will also need to be backwards compatible in the theme to load from both places.